sdk/storage/azure-storage-blob/azure/storage/blob/_shared/encryption.py

```diff
@@ -115,23 +158,244 @@ def __init__(self, content_encryption_IV, encryption_agent, wrapped_content_key,
+
         self.content_encryption_IV = content_encryption_IV
+        self.encrypted_region_info = encrypted_region_info
         self.encryption_agent = encryption_agent
         self.wrapped_content_key = wrapped_content_key
         self.key_wrapping_metadata = key_wrapping_metadata
 
 
-def _generate_encryption_data_dict(kek, cek, iv):
+class GCMBlobEncryptionStream:
+    """
+    A stream that performs AES-GCM encryption on the given data as
+    it's streamed. Data is read and encrypted in regions. The stream
+    will use the same encryption key and will generate a guaranteed unique
+    nonce for each encryption region.
+    """
+    def __init__(
+            self,
+            content_encryption_key: bytes,
+            data_stream: BinaryIO,
+            ):
+        """
+        :param content_encryption_key: The encryption key to use.
+        :param data_stream: The data stream to read data from.
+        """
+        self.content_encryption_key = content_encryption_key
+        self.data_stream = data_stream
+
+        self.offset = 0
+        self.current = b''
+        self.nonce_counter = 0
+
+    def read(self, size: int = -1) -> bytes:
+        """
+        Read data from the stream. Specify -1 to read all available data.
+
+        :param size: The amount of data to read. Defaults to -1 for all data.
+        """
+        result = BytesIO()
+        remaining = sys.maxsize if size == -1 else size
+
+        while remaining > 0:
+            # Start by reading from current
+            if len(self.current) > 0:
+                read = min(remaining, len(self.current))
+                result.write(self.current[:read])
+
+                self.current = self.current[read:]
+                self.offset += read
+                remaining -= read
+
+            if remaining > 0:
+                # Read one region of data and encrypt it
+                data = self.data_stream.read(_GCM_REGION_DATA_LENGTH)
+                if len(data) == 0:
+                    # No more data to read
+                    break
+
+                self.current = self._encrypt_region(data)
+
+        return result.getvalue()
+
+    def _encrypt_region(self, data: bytes) -> bytes:
+        """
+        Encrypt the given region of data using AES-GCM. The result
+        includes the data in the form: nonce + ciphertext + tag.
+        """
+        # Each region MUST use a different nonce
+        nonce = self.nonce_counter.to_bytes(_GCM_NONCE_LENGTH, 'big')
+        self.nonce_counter += 1
+
+        aesgcm = AESGCM(self.content_encryption_key)
+
+        # Returns ciphertext + tag
+        cipertext_with_tag = aesgcm.encrypt(nonce, data, None)
+        return nonce + cipertext_with_tag
+
+
+def is_encryption_v2(encryption_data: Optional[_EncryptionData]) -> bool:
+    """
+    Determine whether the given encryption data signifies version 2.0.
+
+    :param encryption_data: The encryption data. Will return False if this is None.
```

Comment:
Okay this is my last comment on this since it's recurring so I'm sure you get the gist 😛, but I see, is the convention here to list the paramtype above and then omit from docstring? If so then continue with that -- I think I am just unfamiliar with that pattern 😄